File: frontend/src/services/bookingService.ts
import bookingApi from "./bookingApi";
import type { Booking } from "../types/booking";

export const getBookings = async () => {
    const response = await bookingApi.get<Booking[]>("/booking");
    return response.data;
};

export const createBooking = async (booking: Booking) => {
    const response = await bookingApi.post("/booking", booking);
    return response.data;
};

export const updateBooking = async (id: number, booking: Booking) => {
    await bookingApi.put(`/booking/${id}`, booking);
};

export const deleteBooking = async (id: number) => {
    await bookingApi.delete(`/booking/${id}`);
};

export const searchBookings = async (
    date?: string,
    route?: string,
    reference?: string
) => {

    const response = await bookingApi.get(
        "/booking/search",
        {
            params: {
                date,
                route,
                reference
            }
        }
    );


    return response.data;

};

File: backend/TrainTicket.API/Services/BookingService.cs
using TrainTicket.API.Models;
using TrainTicket.API.Repositories;

namespace TrainTicket.API.Services;

public class BookingService
{
    private readonly BookingRepository repository;

    public BookingService(BookingRepository repository)
    {
        this.repository = repository;
    }


    public List<Booking> GetAllBookings()
    {
        return repository.GetAllBookings();
    }


    public Booking? GetBookingById(int id)
    {
        return repository.GetBooking(id);
    }


    public Booking CreateBooking(Booking booking)
    {
        return repository.CreateBooking(booking);
    }


    public void UpdateBooking(Booking booking)
    {
        repository.UpdateBooking(booking);
    }


    public void DeleteBooking(int id)
    {
        repository.DeleteBooking(id);
    }


    public List<Booking> SearchBookings(string? date, string? route, string? reference)
    {
        var bookings = repository.GetAllBookings().AsQueryable();

        if (!string.IsNullOrEmpty(date))
        {
            bookings = bookings.Where(b =>
                b.Schedule.TravelDate.ToString("yyyy-MM-dd").Contains(date)
            );
        }

        if (!string.IsNullOrEmpty(route))
        {
            bookings = bookings.Where(b =>
                b.Route.DepartureStation.Contains(route, StringComparison.OrdinalIgnoreCase) ||
                b.Route.DestinationStation.Contains(route, StringComparison.OrdinalIgnoreCase)
            );
        }

        if (!string.IsNullOrEmpty(reference))
        {
            bookings = bookings.Where(b =>
                b.BookingReference.Contains(reference, StringComparison.OrdinalIgnoreCase)
            );
        }

        return bookings.ToList();
    }

    public decimal GetTotalCost()
    {
        return repository.GetAllBookings().Sum(b => b.TicketPrice);
    }
}